Add WandB user guide (#387)
mila-iqia/mila-docs: docs/userguides/wandb.md

wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
```
</div>

!!! tip
    `uv tool install --upgrade wandb` is a one-time step per cluster. The API
    key is stored in `~/.netrc` after `wandb login`; subsequent jobs do not need
    to re-run `wandb login` unless the key is rotated.

### Offline mode

On clusters without outbound internet access on compute nodes, run in offline
mode and sync after the job completes:

```bash
export WANDB_MODE=offline
srun uv run python main.py
```

After the job finishes, sync the run from the login node:

```bash
wandb sync --sync-all
```

!!! tip
    On DRAC clusters, loading the `httpproxy` module before the `srun` line is
    an alternative to offline mode:
    ```bash
    module load httpproxy
    srun uv run python main.py
    ```

## Initialize and log a training run

Every WandB run starts with `wandb.init()`. Call `wandb.finish()` at the end of
the script — it is called automatically at exit, but an explicit call is safer
in Slurm jobs.

### Initialize a run

```python
import os
import wandb

wandb.init(
    project="wandb-example",                # (1)!
    name=os.environ.get("SLURM_JOB_ID"),    # (2)!

>     resume="allow",                         # (3)!

Reviewer: ```suggestion
    name=os.environ.get("SLURM_JOB_ID"),    # (2)!
    id=os.environ.get("SLURM_JOB_ID"),
    resume="allow",                         # (3)!
```

>     config=vars(args),                      # (4)!

Reviewer: ```suggestion
    config=vars(args) | {f"env/{k}": v for k, v in os.environ.items() if k.startswith("SLURM"},    # (4)!
```

)
```
{ .annotate }

1. Groups runs in the WandB UI. Use one project per research question.
2. Links the run to its log file (`slurm-<JOBID>.out`) for easy
   cross-referencing.
3. Creates a new run if the ID does not exist, or resumes it if it does. Useful
   when combined with checkpointing to recover from preemption.
4. Pass the full `argparse` namespace — every key becomes a searchable,

>    filterable column in the WandB Runs table under **Config**.

Reviewer: ```suggestion
4. Pass the full `argparse` namespace and SLURM environment variables for easier debugging.
    Every key becomes a searchable, filterable column in the WandB Runs table under **Config**.
```

### Log metrics

Call `wandb.log()` inside the batch loop to record training metrics at each
step:

```python
wandb.log(
    {
        "train/accuracy": accuracy,
        "train/loss": loss
    }
)
```

Log validation metrics once per epoch, after the validation loop:

```python
wandb.log(
    {
        "val/accuracy": val_accuracy,
        "val/loss": val_loss,
        "epoch": epoch
    }
)
```

!!! tip
    Prefix metric names with `train/` and `val/`. WandB groups metrics with
    matching prefixes automatically in the Charts panel.

!!! tip "Complete example"
    See the [WandB setup
    example](../examples/good_practices/wandb_setup/index.md) for a complete
    single-GPU training script with WandB logging integrated.

## Organize runs

Three arguments help keep runs organized as a project grows. `name=` and `tags=`
make individual runs easy to identify and filter in the Runs table, while
`group=` clusters related runs — such as multi-seed runs or ablations — under a
single expandable row.

> 
> ```python
> wandb.init(
>     project="wandb-example",
>     name=os.environ.get("SLURM_JOB_ID"),
>     group=os.environ.get("SLURM_ARRAY_JOB_ID", None),   # (1)!
>     resume="allow",
>     tags=["example", "resnet18"],                       # (2)!
>     config=vars(args),
> )

Reviewer: The ID needs to be the same for resuming to work correctly. If there are multiple tasks, then this would might cause an issue, but I don't think we need to worry about that here.
```suggestion
wandb.init(
    project="wandb-example",
    name=os.environ.get("SLURM_JOB_ID"),
    id=os.environ.get("SLURM_JOB_ID"),
    group=os.environ.get("SLURM_ARRAY_JOB_ID"),   # (1)!
    resume="allow",
    tags=["example", "resnet18"],                       # (2)!
    config=vars(args),
)
```

```
{ .annotate }

1. Clusters multi-seed under a single expandable row. Passing

>    `SLURM_ARRAY_JOB_ID` groups every task in a job array together automatically.

Reviewer: ```suggestion
1. Using `SLURM_ARRAY_JOB_ID` automatically as the group clusters all jobs into a single expandable row.
```

2. Labels runs for filtering in the Runs table.

## Diagnose training bottlenecks

WandB records GPU utilization, CPU usage, and memory under the **System** tab of

> every run automatically — no extra code required. These metrics are the first

Reviewer: Les caractères `—` sont des signes de LLMs, personellement je les enlèverais.
```suggestion
every run automatically, no extra code is required. These metrics are the first
```

  I/O-bound.

### Add step timing metrics

GPU utilization alone confirms that a bottleneck exists but does not show where
time is lost. Log data-loading time and compute time separately to quantify each
phase:

```python
import time

# Before the batch loop:
t_step_end = time.perf_counter()

for batch in train_dataloader:
    # The dataloader has already fetched the batch by the time the loop body
    # starts, so the elapsed time since t_step_end is the data-load wait.
    t_data_load_end = time.perf_counter()
    data_load_s = t_data_load_end - t_step_end

    # forward pass, loss computation, backward pass, optimizer step

    t_step_end = time.perf_counter()
    compute_s = t_step_end - t_data_load_end

    wandb.log(
        {
            "train/loss": loss,
            "train/accuracy": accuracy,
            "perf/data_load_s": data_load_s,
            "perf/compute_s": compute_s,
        }
    )
```

When `perf/data_load_s` is consistently larger than `perf/compute_s`, the job is
I/O-bound.

!!! tip
    Common fixes for an I/O bottleneck: increase `num_workers` in the
    `DataLoader`, enable `pin_memory=True`, or copy the dataset to

>     `$SLURM_TMPDIR` before the job starts.

Reviewer: Hmm. C'est pas une mauvaise idée, mais il faudrait aussi ajouter des `torch.cuda.synchronize()` pour faire ça comme il faut, et juste comme ça on risque d'induire un peu les gens en erreur. Je pense que de le faire bien compliquerait beaucoup l'exemple, et ça aurait plus à voir avec le profiling qu'avec wandb.

Je crois qu'il vaudrait mieux s'en tenir à l'identification du problème par le GPU util, et faire un guide plus approfondi sur le profiling eventuellement. Qu'en penses-tu?

Author: Reverted, mais je pense, comme toi aussi, qu'il y a déjà des cuda sync implicite par les f-strings dans les `logger.debug`.

    addresses](https://docs.wandb.ai/guides/app/settings-page/emails). Then log
    out from WandB before proceeding.

1. Go to [wandb.ai](https://wandb.ai) and click **Sign in**.
2. Enter your `@mila.quebec` email address. The password field will disappear
   once a recognized SSO domain is detected.
3. Click **Log in** — the browser will redirect you to the Mila SSO page.
4. Select the **mila.quebec** identity provider. WandB will offer to link the
   existing account to the Mila organization.

### Create a new account

Follow the same SSO steps above. At the account creation prompt, select
**Professional**.

??? question "Which account type to select?"
    Select **Professional** at the account creation prompt. This unlocks team
    features required for the Mila organization. The Mila IT team manages
    organization-level billing, so no personal plan upgrade is required.

## Authenticate the CLI on the cluster

Most WandB Python API calls and sweep commands require a valid API key stored in
the environment.

### Log in interactively

Use this on a login node or an interactive allocation for exploratory work:

```bash

> uvx "wandb@latest" login

Reviewer: ```suggestion
uv tool install wandb
wandb login
```

```
<div class="result" style="border:None; padding:0" markdown>
``` linenums="0"
wandb: Logging into https://api.wandb.ai.
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.
wandb: Paste your API key and hit enter:
wandb: Appending key for api.wandb.ai to your netrc file: /home/mila/u/username/.netrc
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
```
</div>

!!! tip
    The key is stored in `~/.netrc` after `wandb login`. Subsequent jobs on the
    same cluster do not need to re-run this command unless the key is rotated.

### Offline mode

On clusters without outbound internet access on compute nodes, run in offline
mode and sync after the job completes:

```bash
export WANDB_MODE=offline

> srun uv run python main.py

Reviewer: ```suggestion
srun uv run --offline python main.py
```

```

After the job finishes, sync the run from the login node:

```bash

> uvx "wandb@latest" sync --sync-all

Reviewer: ```suggestion
wandb sync --sync-all
```
